flocker/node/agents/cinder.py

```diff
@@ -564,18 +568,27 @@ def detach_volume(self, blockdevice_id):
         )
 
     def destroy_volume(self, blockdevice_id):
+        """
+        Detach Cinder volume identified by blockdevice_id.
+
+        :raises TimeoutException: If the volume is not deleted
+            within the expected time
+        """
         try:
             self.cinder_volume_manager.delete(blockdevice_id)
         except CinderNotFound:
             raise UnknownVolume(blockdevice_id)
-
-        while True:
-            # Don't loop forever here.  FLOC-1853
+        start_time = time.time()
+        # Wait until the volume is not there or until the operation
+        # timesout
+        while(time.time() - start_time < self.timeout):
             try:
                 self.cinder_volume_manager.get(blockdevice_id)
             except CinderNotFound:
-                break
+                return
             time.sleep(1.0)
+        # If the volume is not deleted, raise an exception
+        raise TimeoutException(unicode(blockdevice_id), None, self.timeout)
```

Comment:
The parameters and attributes of `TimeoutException` are undocumented.  Adding a new use of the type is probably a good time to correct that mistake.